File: 525-contiguous-array/525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        
        
        Map<Integer, Integer> map = new  HashMap<>();
        map.put(0,-1);
        int maxlen =0 , cnt =0 ;
        for (int i =0 ; i <nums.length ; i++){
            cnt += (nums[i] ==1 ? 1: -1);
            
            if (map.containsKey(cnt)){
                maxlen = Math.max( maxlen , i-map.get(cnt));
            } else {
                map.put(cnt ,i);
            }
        }
        
        return maxlen; 
        
        
    }
}

File: 525-contiguous-array/525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        
        cnt = maxl = 0
        dic = {0:0}
        for idx, n in enumerate(nums ,1 ):
            
            if n ==0:
                cnt +=1 
            elif n==1:
                cnt -=1 
            
            if cnt in  dic:
                maxl = max(maxl , idx - dic[cnt])
            else:
                dic[cnt] = idx
        return maxl 